[IMP] Add --proxy-trusted-x-for CLI/config option

Yajo opened the pull request:
Without this patch, Odoo is compatible only with one single inverse proxy in front of it.

However, when several proxies are chained in front of it, and all of them use the [proxy protocol][1], Odoo would understand the latest one's IP is the remote IP that is originating the HTTP request.

When several proxies are chained, Werkzeug receives this HTTP header:

    X-Forwarded-For: 1.2.3.4, 5.6.7.8

It means that 1.2.3.4 is the request originator, and 5.6.7.8 is the 1st proxy.

If using [Werkzeug >= 0.14.1][2], which logs the correct IP, the logs would end up with:

    2019-02-14 10:15:09,042 37 INFO prod werkzeug: 5.6.7.8 - - [14/Feb/2019 10:15:09] "GET /web/login HTTP/1.1" 200 -

After this patch, if you use the proper Werkzeug version and configure Odoo with `--proxy-hops=2`, you'd get instead:

    2019-02-14 10:15:09,042 37 INFO prod werkzeug: 1.2.3.4 - - [14/Feb/2019 10:15:09] "GET /web/login HTTP/1.1" 200 -

**Note to migrator:** When migrating to master, you'll probably want to remove `--proxy-mode`, and just use `--proxy-hops` defaulting to `0`. No need to have 2 configs for this, except for backwards compatibility, which is a must here.


[1]: https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
[2]: https://github.com/odoo/odoo/issues/18052#issuecomment-364703091



--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
@Tecnativa

Yajo commented:
@Yenthe666 I summon your pingpacthunder!

![](https://media.giphy.com/media/7ISIRaCMrgFfa/giphy-downsized.gif)

Yenthe666 commented:
I'm totally going places in life now. 🎉 
@d-fence and @odony can I ask you guys to check this please?

odony commented:
Why propose this in 10.0? You guys know the policy ;-)
Moreover, it's best to avoid revealing the network topology in the headers via X-Forwarded-For. And it's brittle, as there are often multiple paths to reach the same origin server, especially with multiple downstream proxies.

A preferred strategy is to have the proxies strip the IPs of downstream proxies they trust. One easy way to do it is with nginx's [realip module](http://nginx.org/en/docs/http/ngx_http_realip_module.html). It can be configured with a list of CIDR ranges, which is quite handy when there is a CDN downstream. There are similar tools for other web servers, and a pure header-rewriting config would be able to accomplish the same, with a bit more work.

Yenthe666 commented:
@odony rebase is done very quickly though :-) +1 for the great insights.

Yajo commented:
> Why propose this in 10.0? You guys know the policy ;-)

The change is fully backwards compatible, and if you like it, I can rebase it or you can always cherry-pick it for master. :sweat_smile: 

> A preferred strategy is to have the proxies strip the IPs of downstream proxies they trust

I agree, although I'm on a situation where I don't control all proxies in the chain, and debugging what's happening is very hard. It seemed like a good improvement to me anyways, given all we have to do is expose an option that was already existing in Werkzeug.

odony commented:
> The change is fully backwards compatible (...) 

We really really really avoid [anything that looks like a new feature/improvement in stable](https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean). Even when it seems safe and backwards compatible. The risk/benefit ratio for such changes is insane. And it's especially true if it touches low-level things like the command line and the framework. It can break deployments in very surprising ways, and it _has_ already happened.

Sure, rebasing is (usually) quick. It just feels nicer when contributors show their understanding of our guidelines, and go the extra mile to make their PR merge-ready, minimizing our manual steps :wink: (like having to rewrite commit messages, to rebase on the correct target and explain why, to request tests, to remove unnecessary chunks in the patch, and so on.)

> I agree, although I'm on a situation where I don't control all proxies in the chain, and debugging what's happening is very hard.

You should only need to control the last proxy in the chain for this to work.

Yajo commented:
Rebased for master.

I changed it to `--proxy-trusted-x-for` to let us be able to sanely expand support when werkzeug 0.15 is released, which will include a refactored `ProxyFix` (see https://github.com/pallets/werkzeug/pull/1314).

Yajo commented:
Any news on this?

Yajo commented:
So? This is for master now... :roll_eyes:

Yajo commented:
Any news? Thanks.

Yenthe666 commented:
Starting to feel like a monologue hmm? @odony can you please give this some TLC again?

odony commented:
I still don't think we need this.

Yajo commented:
It eases deployments. If you don't control any proxy on the chain, you probably needs this. Also, the fact that you don't need this is a good enough reason to block this feature that can help others? :thinking:

odony commented:
We have too many command-line options already, and we should try to remove some, rather than add more, especially for rare corner cases.

From what I understand it mostly seems like you don't know how to configure traefik to do the right thing. There are plenty of alternatives to solve your problem, I'm convinced this should not be the responsibility of the Odoo server in your stack. It's not because a feature exists somewhere in werkzeug (or in any lib we depend on) that it makes sense for us to expose it higher up.

Thanks for your patience and for providing detailed diagnostics of your situation, it will come in handy if someone ever investigates a similar case.
